### PR TITLE
Introduce inline types

### PR DESCRIPTION
**Pull request by mbryzek**
- This change allows user to declare types inline when using api.json format
  - Wherever we declare a type as the name of the model - we can now instead
    declare the type as an object, defining the enum, model or union inline
  - Tests have examples of varying inlining

  - Consider this feature experimental

As one example, this change allows you to define a model
that contains a single field named 'status' that is defined
as an enum named 'user_status'. Instead of having to declare
the enum itself, you can inline the enum into the definition
of the field:
```
    {
      "name": "Api Doc",

      "models": {
        "user": {
          "fields": [
            {
             "name": "status",
              "type": {
                "enum": "user_status",
                "values": [
                  { "name": "guest" },
                  { "name": "registered" }
                ]
              }
            }
          ]
        }
      }
    }
```

API Builder will automatically expand the type, creating the enum - ie. this feature
is to support human editing and does not change in any way the underling service specification.

### CODE REVIEW

File: core/src/test/scala/core/ServiceDefaultsSpec.scala [hunk 1]
```diff
@@ -126,7 +125,7 @@ class ServiceDefaultsSpec extends FunSpec with Matchers {
     buildMinMaxErrors(1, max = Some(2)) should be(Nil)
   }
   
-  it("numeric default ourside of min, max") {
+  it("numeric default ourtide of min, max") {
```
Comment (nemo83) [inline core/src/test/scala/core/ServiceDefaultsSpec.scala]: outside?

**Comment by fabiocognigni** [comment]
Super useful! What should be the designed behaviour for the below edge cases (referring here to enums but valid for models and unions as well) ?
1. same in-lined type name used multiple times but with different values: error message to the user?
2. same in-lined type name and same values used multiple times: merge automatically in one ApiBuilder type?

**Comment by mbryzek** [comment]
In these cases, api builder will continue to return it's error that an enum is defined twice. Related test case at:

https://github.com/apicollective/apibuilder/pull/703/files#diff-8a5daf1dcf7ac77058212dea915ff411R87

**Comment by gheine** [comment]
Nice! 👍
